### ISSUE TITLE
Sentence similarity: Training failed with error "Specified cast is not valid" when using SQL server data.

### ISSUE THREAD

**Issue by v-Hailishi** [dotnet/machinelearning-modelbuilder#2475]
**System Information (please complete the following information):**
Windows OS: Windows-11-Enterprise-22H2
ML.Net Model Builder 2022: 16.14.3.2307701 (Main Build)
Microsoft Visual Studio Enterprise: 2022(17.4.4)
.Net: 6.0

**Describe the bug**
- On which step of the process did you run into an issue: 
Training failed with error "Specified cast is not valid" when using SQL server data.

**TestMatrix**
[home-depot-sentence-similarity-small.csv](https://github.com/dotnet/machinelearning-modelbuilder/files/10534387/home-depot-sentence-similarity-small.csv)

**To Reproduce**
  1. Select Create a new project from the Visual Studio start window.
  2. Choose the C# Console App (.NET Core) project template.
  3. Add model builder by right click on the project.
  4. Select **"Sentence similarity"** scenario.
  5. On Data page, choose the SQL Server data source and select the column for three options.
  6. On the Train page, click "Start training" button to train.
  
**Expected behavior**
Training successful.

**Screenshot**
![image](https://user-images.githubusercontent.com/99375895/216219340-4dbfef9e-c12b-470f-9397-c99860f25529.png)
![image](https://user-images.githubusercontent.com/99375895/216219369-1d79f155-5e96-4c9e-8c96-43afbe7e1341.png)

**Additional context**

**Comment by beccamc** [comment]
@v-Hailishi what is the SQL schema for the table? I want to know the SQL data types.

**Comment by v-Hailishi** [comment]
@beccamc Here are the SQL data types:
<img width="349" alt="image" src="https://user-images.githubusercontent.com/99375895/216896784-f29c9415-c027-4b72-875e-a57088798a46.png">

**Comment by beccamc** [comment]
I am seeing the same problem with the home depot dataset uploaded to SQL. Here are the column types... 

![image](https://user-images.githubusercontent.com/9122518/217070109-69532327-4a3a-460d-bcca-679c3473dd2b.png)

**Comment by beccamc** [comment]
@LittleLittleCloud I'm looking for some help here... 
- DB command string has correct casts 
- Column options have everything set correctly. 
- I can't reproduce the problem in our generated code when I load the IDataView from the database. 

Why are we even calling into the Database Loader from the `SentenceSimilarityTrainer.CreateTrainer` code path? We're passing in the IDataView, why does it go back to the database loader? 
```

Specified cast is not valid.
   at System.Data.SqlClient.SqlBuffer.get_Single()
   at Microsoft.ML.Data.DatabaseLoader.Cursor.<>c__DisplayClass37_0.<CreateSingleGetterDelegate>b__0(Single& value)
   at Microsoft.ML.Data.ColumnCursorExtensions.<GetColumnDirect>d__2`1.MoveNext()
   at Microsoft.ML.TorchSharp.NasBert.SentenceSimilarityTrainer.Trainer.GetRowCountAndSetLabelCount(IDataView input)
   at Microsoft.ML.TorchSharp.NasBert.NasBertTrainer`2.TrainerBase..ctor(NasBertTrainer`2 parent, IChannel ch, IDataView input)
   at Microsoft.ML.TorchSharp.NasBert.SentenceSimilarityTrainer.CreateTrainer(NasBertTrainer`2 parent, IChannel ch, IDataView input)
   at Microsoft.ML.TorchSharp.NasBert.NasBertTrainer`2.Fit(IDataView input)
   at Microsoft.ML.Data.EstimatorChain`1.Fit(IDataView input)
   at Microsoft.ML.AutoML.SweepablePipelineRunner.Run(TrialSettings settings)
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.ML.AutoML.AutoMLExperiment.<RunAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.ML.ModelBuilder.AutoMLService.Experiments.SentenceSimilarityExperiment.<ExecuteAsync>d__13.MoveNext() in C:\Repos\machinelearning-tools\src\Microsoft.ML.ModelBuilder.AutoMLService\Experiments\SentenceSimilarityExperiment.cs:line 106
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.ML.ModelBuilder.AutoMLEngine.<StartTrainingAsync>d__21.MoveNext() in C:\Repos\machinelearning-tools\src\Microsoft.ML.ModelBuilder.AutoMLService\AutoMLEngineService\AutoMLEngine.cs:line 145
```

**Comment by LittleLittleCloud** [comment]
@beccamc 
`IDataView` loads data in a lazy way, that means it won't start loading data until that column is needed.

> I can't reproduce the problem in our generated code when I load the IDataView from the database.

Can you try calling
```
data.GetColumn<float>("relevance").ToArray()
```
This should trigger the error.


And I wonder that SQL float is actually double in C#? You should be able to verify that by changing float to double in the upper code.

**Comment by beccamc** [comment]
Problem here was that the index was incorrect! Basically we were pulling the wrong column with ignored values. This is an existing bug and wasn't introduced in the Sentence Similarity scenario.

**Comment by v-Hailishi** [comment]
The bug has been fixed on the latest main build 17.14.3.2311302.
<img width="325" alt="image" src="https://user-images.githubusercontent.com/99375895/218655445-3496b775-e845-43b5-8603-44f8607415f1.png">
<img width="314" alt="image" src="https://user-images.githubusercontent.com/99375895/218655493-62ece394-063c-4f00-8c63-4a77c56eea77.png">
